modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -389,10 +408,19 @@ public Checkpointer getCheckpointer() {
             final FileLockHolder preLocked = kernalCtx.pdsFolderResolver()
                 .resolveFolders()
                 .getLockedFileLockHolder();
+
             if (preLocked == null)
                 fileLockHolder = new FileLockHolder(storeMgr.workDir().getPath(), kernalCtx, log);
 
             persStoreMetrics.wal(cctx.wal());
+
+            recoveryIo = new RecoveryIoImp(kernalCtx, log);
```

Comment:
RecoveryIo looks redundant, we do not use such a pattern. Instead, we probably should add a recovery processor which will add this listener on startup and manager recovery future(s).